### ISSUE TITLE
recursive structure AndQuestion 

### ISSUE THREAD

**Issue by ashu2012** [japgolly/learning#1]
https://github.com/japgolly/learning/blob/461bb0c5a75039533d96542c35ac0b5ee4081430/talks/20180827-MelbCompose-Quantifiers/existentials.scala#L21

how this recursive structure is used , is not clear from video. could you explain in simpler terms. thanks

final type AndQuestion = SurveyType.AndQuestion { val surveyType: self.type }
final def AndQuestion(q: Question): self.AndQuestion =
new SurveyType.AndQuestion {
override val surveyType: self.type = self
override val question = q
}

**Comment by ashu2012** [comment]
trait SurveyType mention object companion object SurveyType which implements another trait AndQuestion has reference of trait SurveyType inside that .
didn't get this part and intuition behind this from Youtube video thanks. Thanks for your awesome lib scalajs-react

**Comment by japgolly** [comment]
It's not recursive. So in `object SurveyType` there's a `trait AndQuestion` which a new data type definition that has a `val surveyType`. Then inside `trait SurveyType` we create a type alias (so not a new type) that is a shortcut to the `AndQuestion` in the `object`, but this time we tell the compiler that we know more information about the `val surveyType` type, namely that it matches the type of the outer `trait SurveyType`.

This means `SurveyType.AndQuestion` can be any SurveyType and we can also have what look to be subclasses like `FreeText.AndQuestion`. Really `FreeText.AndQuestion` is the same as saying `SurveyType.AndQuestion { val surveyType: FreeText }` which is quite verbose and doesn't read nicely.
